Feat: implement url shortener service

DimaOsinnii opened the pull request:
Tasks:
- create basic layout using picocss
- implement form with svelte enhance for creating short url
- integrate cloudflare Pages & KV
- implement collection request data during redirecting from short url
- create short url stats page
- create list of all created short urls

DimaOsinnii committed:
create basic app layout

DimaOsinnii committed:
create links and history services; create base errors; create validation helpers; create app constants

DimaOsinnii committed:
implement home page with main form and server action for creating short urls

DimaOsinnii committed:
create api endpoint for redirecting from short to long url and saving required data from request in history kv

DimaOsinnii committed:
implement short url stat page

DimaOsinnii committed:
implement list of all created urls page

DimaOsinnii committed:
fix wrangler

DimaOsinnii committed:
hotfix

DimaOsinnii committed:
hotfix

DimaOsinnii committed:
hotfix

DimaOsinnii committed:
improvements

DimaOsinnii committed:
remove wrangler.toml

DimaOsinnii committed:
fix styles

nosovk commented on src/routes/+page.server.ts:
```diff
@@ -0,0 +1,43 @@
+		const origin = new URL(request.url).origin;
+
+		const url = String(formData.get(FIELD_NAMES.URL));
+		const shortUrl = String(formData.get(FIELD_NAMES.SHORT_URL));
```
shouldn't we add fileguard to check, isn't it file inside formdata?

nosovk commented on src/routes/[shortUrl]/stats/+page.svelte:
```diff
@@ -0,0 +1,57 @@
+<script lang="ts">
```
there should be invalidation on reload, I think its not nice to cache stats page (you make few clicks, go to stats page and its empty)

DimaOsinnii replied:
I do invalidateAll in onMount on each dynamic data page - /shortUrl/stats and /list but nothing changed. I also check without cache in the browser and from  incognito tabs.  I think the problem is in Cloudflare KV,  the latency  from write to read is too big, and i read few post and KV docs where people said that it can be happed, maybe because of hot and cold reads.

So i think maybe Cloudflare KV is not perfect match for this kind of task or maybe it need to be used as a cache, not storage

https://developers.cloudflare.com/kv/reference/how-kv-works/
https://community.cloudflare.com/t/delay-in-workers-kv/149463

nosovk commented on src/routes/[shortUrl]/stats/+page.server.ts:
```diff
@@ -0,0 +1,18 @@
+import { createHistoryService } from '../../../services/history';
```
better use "$lib"

nosovk commented on src/constants.ts:
```diff
@@ -0,0 +1,19 @@
+};
+
+export const FIELD_LABELS = {
+	[FIELD_NAMES.URL]: 'URL',
```
я понимаю, что хотелось сделать общий ключи между фронтом и "беком" свелта, но почему не использовать 
'string1'|'string2'? На клиенте сейчас оно выглядит как куча интерполяций, так можно было бы избежать такого

nosovk commented on src/routes/+page.server.ts:
```diff
@@ -0,0 +1,43 @@
+				errors[FIELD_NAMES.SHORT_URL] = error.message;
+			} else if (error instanceof InvalidUrlError) {
+				errors[FIELD_NAMES.URL] = error.message;
+			} else if (error instanceof InvalidShortUrlError) {
```
probably we shouldn't use elseif here?

nosovk commented:
try to put as url:
```
data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAApgAAAKYB3X3/OAAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAANCSURBVEiJtZZPbBtFFMZ/M7ubXdtdb1xSFyeilBapySVU8h8OoFaooFSqiihIVIpQBKci6KEg9Q6H9kovIHoCIVQJJCKE1ENFjnAgcaSGC6rEnxBwA04Tx43t2FnvDAfjkNibxgHxnWb2e/u992bee7tCa00YFsffekFY+nUzFtjW0LrvjRXrCDIAaPLlW0nHL0SsZtVoaF98mLrx3pdhOqLtYPHChahZcYYO7KvPFxvRl5XPp1sN3adWiD1ZAqD6XYK1b/dvE5IWryTt2udLFedwc1+9kLp+vbbpoDh+6TklxBeAi9TL0taeWpdmZzQDry0AcO+jQ12RyohqqoYoo8RDwJrU+qXkjWtfi8Xxt58BdQuwQs9qC/afLwCw8tnQbqYAPsgxE1S6F3EAIXux2oQFKm0ihMsOF71dHYx+f3NND68ghCu1YIoePPQN1pGRABkJ6Bus96CutRZMydTl+TvuiRW1m3n0eDl0vRPcEysqdXn+jsQPsrHMquGeXEaY4Yk4wxWcY5V/9scqOMOVUFthatyTy8QyqwZ+kDURKoMWxNKr2EeqVKcTNOajqKoBgOE28U4tdQl5p5bwCw7BWquaZSzAPlwjlithJtp3pTImSqQRrb2Z8PHGigD4RZuNX6JYj6wj7O4TFLbCO/Mn/m8R+h6rYSUb3ekokRY6f/YukArN979jcW+V/S8g0eT/N3VN3kTqWbQ428m9/8k0P/1aIhF36PccEl6EhOcAUCrXKZXXWS3XKd2vc/TRBG9O5ELC17MmWubD2nKhUKZa26Ba2+D3P+4/MNCFwg59oWVeYhkzgN/JDR8deKBoD7Y+ljEjGZ0sosXVTvbc6RHirr2reNy1OXd6pJsQ+gqjk8VWFYmHrwBzW/n+uMPFiRwHB2I7ih8ciHFxIkd/3Omk5tCDV1t+2nNu5sxxpDFNx+huNhVT3/zMDz8usXC3ddaHBj1GHj/As08fwTS7Kt1HBTmyN29vdwAw+/wbwLVOJ3uAD1wi/dUH7Qei66PfyuRj4Ik9is+hglfbkbfR3cnZm7chlUWLdwmprtCohX4HUtlOcQjLYCu+fzGJH2QRKvP3UNz8bWk1qMxjGTOMThZ3kvgLI5AzFfo379UAAAAASUVORK5CYII=
```

nosovk commented on src/lib/index.ts:
```diff
@@ -1 +1,16 @@
 // place files you want to import through the `$lib` alias in this folder.
+
+//TODO: make more stronger validation, if its required from domain logic
```
actually svelt makes validation to filter file://

DimaOsinnii committed:
clean up

refactor form action, page

remove unnecessary errors; create validation for create url form data; remove field configs instead create union type

DimaOsinnii committed:
add invalidation

DimaOsinnii committed:
fix invalidation

DimaOsinnii commented:
> try to put as url:
> 
> ```
> data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAApgAAAKYB3X3/OAAAABl0RVh0U29mdHdhcmUAd3d3Lmlua3NjYXBlLm9yZ5vuPBoAAANCSURBVEiJtZZPbBtFFMZ/M7ubXdtdb1xSFyeilBapySVU8h8OoFaooFSqiihIVIpQBKci6KEg9Q6H9kovIHoCIVQJJCKE1ENFjnAgcaSGC6rEnxBwA04Tx43t2FnvDAfjkNibxgHxnWb2e/u992bee7tCa00YFsffekFY+nUzFtjW0LrvjRXrCDIAaPLlW0nHL0SsZtVoaF98mLrx3pdhOqLtYPHChahZcYYO7KvPFxvRl5XPp1sN3adWiD1ZAqD6XYK1b/dvE5IWryTt2udLFedwc1+9kLp+vbbpoDh+6TklxBeAi9TL0taeWpdmZzQDry0AcO+jQ12RyohqqoYoo8RDwJrU+qXkjWtfi8Xxt58BdQuwQs9qC/afLwCw8tnQbqYAPsgxE1S6F3EAIXux2oQFKm0ihMsOF71dHYx+f3NND68ghCu1YIoePPQN1pGRABkJ6Bus96CutRZMydTl+TvuiRW1m3n0eDl0vRPcEysqdXn+jsQPsrHMquGeXEaY4Yk4wxWcY5V/9scqOMOVUFthatyTy8QyqwZ+kDURKoMWxNKr2EeqVKcTNOajqKoBgOE28U4tdQl5p5bwCw7BWquaZSzAPlwjlithJtp3pTImSqQRrb2Z8PHGigD4RZuNX6JYj6wj7O4TFLbCO/Mn/m8R+h6rYSUb3ekokRY6f/YukArN979jcW+V/S8g0eT/N3VN3kTqWbQ428m9/8k0P/1aIhF36PccEl6EhOcAUCrXKZXXWS3XKd2vc/TRBG9O5ELC17MmWubD2nKhUKZa26Ba2+D3P+4/MNCFwg59oWVeYhkzgN/JDR8deKBoD7Y+ljEjGZ0sosXVTvbc6RHirr2reNy1OXd6pJsQ+gqjk8VWFYmHrwBzW/n+uMPFiRwHB2I7ih8ciHFxIkd/3Omk5tCDV1t+2nNu5sxxpDFNx+huNhVT3/zMDz8usXC3ddaHBj1GHj/As08fwTS7Kt1HBTmyN29vdwAw+/wbwLVOJ3uAD1wi/dUH7Qei66PfyuRj4Ik9is+hglfbkbfR3cnZm7chlUWLdwmprtCohX4HUtlOcQjLYCu+fzGJH2QRKvP3UNz8bWk1qMxjGTOMThZ3kvgLI5AzFfo379UAAAAASUVORK5CYII=
> ```

DimaOsinnii commented on src/lib/validation.ts:
```diff
@@ -0,0 +1,35 @@
+
+	let parsedUrl: URL;
+
+	try {
+		parsedUrl = new URL(url);
+	} catch (_) {
+		throw new CreateUrlError('url', 'Invalid URL');
+	}
+
+	if (!allowedProtocols.includes(parsedUrl.protocol)) {
+		throw new CreateUrlError('url', 'Only HTTP and HTTPS URLs are allowed');
+	}
```
I read about URL.canParse method, but i need also check protocol of the actual url and canParse does not allowe this, so the solution was bettwen native api and regexp i choose first approach
